subsys/net/ip/net_if.c

```diff
@@ -455,10 +456,24 @@ enum net_verdict net_if_send_data(struct net_if *iface, struct net_pkt *pkt)
 		goto done;
 	}
 
-	if (IS_ENABLED(CONFIG_NET_OFFLOAD) && !net_if_l2(iface)) {
-		NET_WARN("no l2 for iface %p, discard pkt", iface);
-		verdict = NET_DROP;
-		goto done;
+	/* The check for CONFIG_NET_*_OFFLOAD here is an optimization;
+	 * This is currently the only way for net_if_l2 to be NULL or missing send().
+	 */
+	if (IS_ENABLED(CONFIG_NET_OFFLOAD) || IS_ENABLED(CONFIG_NET_SOCKETS_OFFLOAD)) {
+		l2 = net_if_l2(iface);
+		if (!l2) {
+			/* Offloaded ifaces may choose not to use an L2 at all. */
+			NET_WARN("no l2 for iface %p, discard pkt", iface);
+			verdict = NET_DROP;
+			goto done;
+		} else if (!l2->send) {
```

Comment:
Prefer explicit checks over implicit

```suggestion
		if (l2 == NULL) {
			/* Offloaded ifaces may choose not to use an L2 at all. */
			NET_WARN("no l2 for iface %p, discard pkt", iface);
			verdict = NET_DROP;
			goto done;
		} else if (l2->send == NULL) {
```